File: Sources/SwiftAria2/handle.swift
import CxxAria2
import Foundation

public final class DownloadHandle {
    package typealias CompletionHandler = (URL?, URLResponse?, (any Error)?) -> Void

    package var completionHandler: CompletionHandler?

    let gid: aria2.A2Gid

    private let handle: CxxAria2.DownloadHandle

    init(session: Session, gid: aria2.A2Gid) {
        self.gid = gid
        let ariaHandle = aria2.getDownloadHandle(session.sessionPtr, gid)
        self.handle = CxxAria2.DownloadHandle(ariaHandle)
    }
}

public extension DownloadHandle {
    var status: aria2.DownloadStatus {
        handle.getStatus()
    }

    var dir: String {
        String(handle.getDir())
    }

    var errorCode: Int32 {
        handle.getErrorCode()
    }

    var files: [aria2.FileData] {
        Array(handle.getFiles())
    }

    var filesPaths: [String] {
        handle.getFiles().map { String($0.path) }
    }

    var downloadSpeed: Int32 {
        handle.getDownloadSpeed()
    }

    var uploadSpeed: Int32 {
        handle.getUploadSpeed()
    }

    var totalLength: Int64 {
        handle.getTotalLength()
    }

    var completedLength: Int64 {
        handle.getCompletedLength()
    }
}

File: Sources/CxxAria2/handle.cpp
#include "handle.h"

using namespace CxxAria2;

DownloadHandle::DownloadHandle(aria2::DownloadHandle* h) {
    this->handle = h;
}

aria2::DownloadStatus DownloadHandle::getStatus() const {
    return this->handle->getStatus();
}

int64_t DownloadHandle::getTotalLength() const {
    return this->handle->getTotalLength();
}

int64_t DownloadHandle::getCompletedLength() const {
    return this->handle->getCompletedLength();
}

int64_t DownloadHandle::getUploadLength() const {
    return this->handle->getUploadLength();
}

std::string DownloadHandle::getBitfield() const {
    return this->handle->getBitfield();
}

int DownloadHandle::getDownloadSpeed() const {
    return this->handle->getDownloadSpeed();
}

int DownloadHandle::getUploadSpeed() const {
    return this->handle->getUploadSpeed();
}

const std::string& DownloadHandle::getInfoHash() const {
    return this->handle->getInfoHash();
}

std::size_t DownloadHandle::getPieceLength() const {
    return this->handle->getPieceLength();
}

int DownloadHandle::getNumPieces() const {
    return this->handle->getNumPieces();
}

int DownloadHandle::getConnections() const {
    return this->handle->getConnections();
}

int DownloadHandle::getErrorCode() const {
    return this->handle->getErrorCode();
}

const std::vector<aria2::A2Gid>& DownloadHandle::getFollowedBy() const {
    return this->handle->getFollowedBy();
}

aria2::A2Gid DownloadHandle::getFollowing() const {
    return this->handle->getFollowing();
}

aria2::A2Gid DownloadHandle::getBelongsTo() const {
    return this->handle->getBelongsTo();
}

const std::string DownloadHandle::getDir() const {
    return this->handle->getDir();
}

const std::vector<aria2::FileData> DownloadHandle::getFiles() const {
    return this->handle->getFiles();
}

int DownloadHandle::getNumFiles() const {
    return this->handle->getNumFiles();
}

aria2::FileData DownloadHandle::getFile(int index) const {
    return this->handle->getFile(index);
}

aria2::BtMetaInfoData DownloadHandle::getBtMetaInfo() const {
    return this->handle->getBtMetaInfo();
}

const std::string& DownloadHandle::getOption(const std::string& name) const {
    return this->handle->getOption(name);
}

aria2::KeyVals DownloadHandle::getOptions() const {
    return this->handle->getOptions();
}
